### ISSUE TITLE
CrossEntropy backward: gradient missing 1/batch_size factor

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#90]
**Audit context:** Filed as part of a USERAPI audit of ODT (Pass 1, April 2026).
Line numbers below reference `es-ude/OnDeviceTraining@b0a737a9d1ff71d71187c6600b56ed5eff4eaab0`.

## Summary

`crossEntropySoftmaxBackwardFloat` computes the gradient without dividing by `batch_size`. When using CE with SGD, the gradient magnitude flowing into the optimizer is exactly `batch_size ×` PyTorch's `reduction='mean'` gradient. Users who tune LR by PyTorch convention (common when porting) work with a silently-boosted effective LR of `batch_size × user_lr`.

## Severity

| Axis | Value |
|---|---|
| Priority | High |
| Affected users | All CE + SGD users (the canonical classification setup) |
| Failure mode | Silent |
| Blast radius | Training correctness |

## Root cause

`src/src/loss_functions/CrossEntropy.c:57-67` — https://github.com/es-ude/OnDeviceTraining/blob/b0a737a9d1ff71d71187c6600b56ed5eff4eaab0/src/src/loss_functions/CrossEntropy.c#L57-L67

```c
for (size_t i = 0; i < batchSize * numClasses; i++) {
    lossGrad[i] = softmaxOutputFloat[i] - distributionFloat[i];
}
```

This is the raw `softmax(x) − target` gradient, computed element-wise over the whole flat `batchSize × numClasses` buffer. PyTorch's default `F.nll_loss(..., reduction='mean')` divides the summed gradient by `batch_size`. ODT does not, so the gradient magnitude passed to SGD is exactly `batch_size` times PyTorch's mean-reduction gradient.

The forward-loss path is correct and per-sample — `TrainingBatchDefault.c:18` divides `totalLoss` by `batch->size`, and `TrainingEpochDefault.c:25` divides by `numberOfBatches`. So reported `train_loss` and `eval_loss` share units. The mismatch lives only in the backward / gradient path.

## Reproducer

Weight-aligned state-dump comparison: 784→32→10 MLP, BS=32, bitwise-identical initial weights loaded into both frameworks, first 32 unshuffled MNIST samples driving both. After one forward+backward:

| Comparison | `max_abs_diff` (across all 4 parameter grads) | `mean_abs_diff` |
|---|---|---|
| ODT gradient vs PyTorch **sum**-reduction | ≤ 1.00e-5 | ≈ 6e-7 |
| (ODT × 1/32) vs PyTorch **mean**-reduction | ≤ 1.5e-8 | ≈ 1e-9 |

Tensors compared: `post_grad_w_0`, `post_grad_b_0`, `post_grad_w_1`, `post_grad_b_1`. Agreement with PyTorch's sum-reduction is at FP32 accumulation-order precision. The factor is exactly `batch_size = 32`.

## Impact

- Cross-framework LR tuning is silently off by a factor of `batch_size`.
- At the audit configuration (MNIST, BS=32), effective LR is `32 × user_lr`. A PyTorch-tutorial `lr=0.001` actually trains at `0.032`.
- Depth-dependent dynamics observed in the audit: at `depth=1` the 32× boost happens to help (baseline `lr=0.001` was under-trained, PyTorch at `lr=0.032` hits ~96%, ODT at effective 0.032 hits ~87%). At `depth=4+`, a 32×-too-large LR becomes plausibly destabilizing — consistent with the observed 3.6× gap widening between 2-layer and 5-layer MLPs in the test suite.
- MCU configurations typically use `BS=1`, where the factor degenerates to `1×`, hiding the bug entirely on embedded targets and making it invisible during on-device validation.

## Recommended fix

**Option A (recommended):** divide by `batchSize` inside the kernel. Zero API-surface change; one line.

```c
for (size_t i = 0; i < batchSize * numClasses; i++) {
    lossGrad[i] = (softmaxOutputFloat[i] - distributionFloat[i]) / (float)batchSize;
}
```

**Option B:** leave the code; document the LR convention in `TrainingLoopApi.h` ("effective LR is `batch_size × user_lr` when using `crossEntropySoftmax`"). Less friendly for PyTorch porters and still silent for anyone who does not read the header.

Option A is strongly recommended because it aligns ODT's gradient semantics with PyTorch's default (the framework most likely to be the cross-check reference).

## Verification

Unit test: construct `lossGrad`, `softmaxOutput`, `distribution` tensors with `batchSize=32`, `numClasses=10`, `lossGrad` zeroed, fixed arbitrary values in the other two. Call `crossEntropySoftmaxBackwardFloat`. Assert element-wise:

```c
for (size_t i = 0; i < 32 * 10; i++) {
    float expected = (softmaxOutput[i] - distribution[i]) / 32.0f;
    assert(fabsf(lossGrad[i] - expected) < 1e-6f);
}
```

Before fix: fails with values 32× larger than expected. After fix: passes.

Optional integration check: re-run any existing CE+SGD training test; training-loss trajectory will change numerically (not just shift), so test expectations will need updating. Classification accuracy should stay within seed noise once the test's `lr` is re-tuned by dividing by `batch_size`.

## Related

No existing upstream issues are directly related. Cross-references for this audit batch live in the audit tracking index issue.

**Comment by LeoBuron** [comment]
Fixed on `develop` in 28088b3 (`fix(#90): add lossConfig_t with reduction modes; CE/MSE backward respect MEAN/SUM`). Closing now since the commit is already on `develop`.
